GCP Workload Identity problem

**Describe the bug**
All necessary resources are created, but OpenCost still requires a key.

**To Reproduce**
Create all necessary resources (gsa, workload identity, policy:

```
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMServiceAccount
metadata:
  name: opencost
  namespace: victoria-metrics
spec:
  displayName: GCP SA bound to K8S SA project-id[opencost]
  resourceID: opencost
---
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicyMember
metadata:
  name: opencost-workload
  namespace: victoria-metrics
spec:
  member: serviceAccount:project-id.svc.id.goog[victoria-metrics/opencost]
  resourceRef:
    apiVersion: iam.cnrm.cloud.google.com/v1beta1
    kind: IAMServiceAccount
    name: opencost
  role: roles/iam.workloadIdentityUser
---
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPolicyMember
metadata:
  name: opencost-bigquery-data-viewer
  namespace: victoria-metrics
spec:
  member: serviceAccount:opencost@project-id.iam.gserviceaccount.com
  resourceRef:
    apiVersion: resourcemanager.cnrm.cloud.google.com/v1beta1
    external: projects/project-id
    kind: Project
  role: dataViewer, etc...
```

**Expected behavior**
As I understand, OpenCost should start working without requiring a key.

**Which version of OpenCost are you using?**
Version: 1.114.0 (47c4b68)

**Logs**
```
2025-05-05T10:07:21.667826433Z ??? Log level set to info
2025-05-05T10:07:21.667926493Z INF Starting cost-model version 1.114.0 (47c4b68)
2025-05-05T10:07:21.667935913Z INF Kubernetes enabled: true
2025-05-05T10:07:21.667989743Z INF Prometheus/Thanos Client Max Concurrency set to 5
2025-05-05T10:07:21.677586122Z INF Success: retrieved the 'up' query against prometheus at: http://vmselect-victoria-metrics.victoria-metrics:8481/select/0/prometheus
2025-05-05T10:07:21.678106042Z INF Retrieved a prometheus config file from: http://vmselect-victoria-metrics.victoria-metrics:8481/select/0/prometheus
2025-05-05T10:07:21.678503682Z INF Using scrape interval of 60.000000
2025-05-05T10:07:21.740145943Z INF Found ProviderID starting with "gce", using GCP Provider
panic: Supply a GCP Key to start getting data

goroutine 1 [running]:
github.com/opencost/opencost/pkg/costmodel.Initialize(0xc00283fe60, {0x0, 0x0, 0x0})
	/home/runner/work/opencost/opencost/opencost/pkg/costmodel/router.go:1265 +0x29de
github.com/opencost/opencost/pkg/cmd/costmodel.Execute(0x27b7301?)
	/home/runner/work/opencost/opencost/opencost/pkg/cmd/costmodel/costmodel.go:46 +0x1ce
github.com/opencost/opencost/pkg/cmd.Execute.newCostModelCommand.func1(0xc000d10f00?, {0x303c3a1?, 0x4?, 0x303c3a5?})
	/home/runner/work/opencost/opencost/opencost/pkg/cmd/commands.go:108 +0x2f
github.com/spf13/cobra.(*Command).execute(0xc00134adc8, {0x5599820, 0x0, 0x0})
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856 +0x68a
github.com/spf13/cobra.(*Command).ExecuteC(0xc00134b088)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x38d
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
github.com/opencost/opencost/pkg/cmd.Execute(0x68?, {0x0, 0x0, 0x0})
	/home/runner/work/opencost/opencost/opencost/pkg/cmd/commands.go:61 +0x3a5
main.main()
	/home/runner/work/opencost/opencost/opencost/cmd/costmodel/main.go:11 +0x1c
[gke_project-id_europe-central2-c_project-id-eu-c2-main-0|victoria-metrics] ➜ 
```

**Helm Chart values**
```
  fullnameOverride: "opencost"
  serviceAccount:
    annotations:
      iam.gke.io/gcp-service-account: opencost@project-id.iam.gserviceaccount.com
  opencost:
    ui:
      enabled: false
    cloudCost:
      enabled: true
    metrics:
      serviceMonitor:
        enabled: true
    prometheus:
      external:
        enabled: true
        url: http://vmselect-victoria-metrics.victoria-metrics:8481/select/0/prometheus
      internal:
        enabled: false
```

We are experiencing the same exact problem. Even though workload identity is created, and properly bound, It panics at the start if key is not given:
```
panic: Supply a GCP Key to start getting data
```
And, if you are to give a dummy key so it does not panic, it will load that dummy key and give another error:
```
...
"message": "API key not valid. Please pass a valid API key."
```

--- a/pkg/cloud/gcp/provider.go
+++ b/pkg/cloud/gcp/provider.go
@@ -1373,9 +1373,12 @@ func (gcp *GCP) getReservedInstances() ([]*GCPReservedInstance, error) {
 		return nil, err
 	}
 
-	projID, err := gcp.MetadataClient.ProjectID()
-	if err != nil {
-		return nil, err
+	projID := gcp.ProjectID
+	if projID == "" {
+		projID, err = gcp.MetadataClient.ProjectID()
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	commitments, err := computeService.RegionCommitments.AggregatedList(projID).Do()

--- a/pkg/cloud/gcp/provider.go
+++ b/pkg/cloud/gcp/provider.go
@@ -39,6 +39,7 @@ import (
 const GKE_GPU_TAG = "cloud.google.com/gke-accelerator"
 const BigqueryUpdateType = "bigqueryupdate"
 const BillingAPIURL = "https://cloudbilling.googleapis.com/v1/services/6F81-5844-456A/skus"
+const GCPCloudOAuthScope = "https://www.googleapis.com/auth/cloud-platform"
 
 const (
 	GCPHourlyPublicIPCost = 0.01
@@ -981,7 +982,7 @@ func (gcp *GCP) getBillingAPIClientAndURL(apiKey, currencyCode string) (*http.Cl
 		return http.DefaultClient, url.String(), nil
 	}
 
-	googleHttpClient, err := google.DefaultClient(context.TODO(), "https://www.googleapis.com/auth/cloud-platform")
+	googleHttpClient, err := google.DefaultClient(context.TODO(), GCPCloudOAuthScope)
 	if err != nil {
 		log.Errorf("GCP Billing API: Workload Identity detected but failed to create authenticated client: %v", err)
 		return nil, "", err

--- a/pkg/cloud/gcp/provider.go
+++ b/pkg/cloud/gcp/provider.go
@@ -1373,7 +1373,12 @@ func (gcp *GCP) getReservedInstances() ([]*GCPReservedInstance, error) {
 		return nil, err
 	}
 
-	commitments, err := computeService.RegionCommitments.AggregatedList(gcp.ProjectID).Do()
+	projID, err := gcp.MetadataClient.ProjectID()
+	if err != nil {
+		return nil, err
+	}
+
+	commitments, err := computeService.RegionCommitments.AggregatedList(projID).Do()
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/cloud/gcp/provider.go
+++ b/pkg/cloud/gcp/provider.go
@@ -6,6 +6,7 @@ import (
 	"io"
 	"math"
 	"net/http"
+	"net/url"
 	"os"
 	"path"
 	"regexp"
@@ -37,7 +38,7 @@ import (
 
 const GKE_GPU_TAG = "cloud.google.com/gke-accelerator"
 const BigqueryUpdateType = "bigqueryupdate"
-const BillingAPIURLFmt = "https://cloudbilling.googleapis.com/v1/services/6F81-5844-456A/skus?key=%s&currencyCode=%s"
+const BillingAPIURL = "https://cloudbilling.googleapis.com/v1/services/6F81-5844-456A/skus"
 
 const (
 	GCPHourlyPublicIPCost = 0.01
@@ -955,8 +956,38 @@ func (gcp *GCP) parsePage(r io.Reader, inputKeys map[string]models.Key, pvKeys m
 	return gcpPricingList, nextPageToken, nil
 }
 
-func (gcp *GCP) getBillingAPIURL(apiKey, currencyCode string) string {
-	return fmt.Sprintf(BillingAPIURLFmt, apiKey, currencyCode)
+func (gcp *GCP) buildBillingAPIURL(apiKey, currencyCode string) *url.URL {
+	url, err := url.Parse(BillingAPIURL)
+	if err != nil {
+		panic("BillingAPIURL must be a valid URL")
+	}
+
+	query := url.Query()
+	query.Add("currencyCode", currencyCode)
+
+	if apiKey != "" {
+		query.Add("key", apiKey)
+	}
+
+	url.RawQuery = query.Encode()
+
+	return url
+}
+
+func (gcp *GCP) getBillingAPIClientAndURL(apiKey, currencyCode string) (*http.Client, string, error) {
+	url := gcp.buildBillingAPIURL(apiKey, currencyCode)
+
+	if apiKey != "" {
+		return http.DefaultClient, url.String(), nil
+	}
+
+	googleHttpClient, err := google.DefaultClient(context.TODO(), "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		log.Errorf("GCP Billing API: Workload Identity detected but failed to create authenticated client: %v", err)
+		return nil, "", err
+	}
+
+	return googleHttpClient, url.String(), nil
 }
 
 func (gcp *GCP) parsePages(inputKeys map[string]models.Key, pvKeys map[string]models.PVKey) (map[string]*GCPPricing, error) {
@@ -966,7 +997,10 @@ func (gcp *GCP) parsePages(inputKeys map[string]models.Key, pvKeys map[string]mo
 		return nil, err
 	}
 
-	url := gcp.getBillingAPIURL(gcp.APIKey, c.CurrencyCode)
+	httpClient, url, err := gcp.getBillingAPIClientAndURL(gcp.APIKey, c.CurrencyCode)
+	if err != nil {
+		return nil, err
+	}
 
 	var parsePagesHelper func(string) error
 	parsePagesHelper = func(pageToken string) error {
@@ -975,7 +1009,7 @@ func (gcp *GCP) parsePages(inputKeys map[string]models.Key, pvKeys map[string]mo
 		} else if pageToken != "" {
 			url = url + "&pageToken=" + pageToken
 		}
-		resp, err := http.Get(url)
+		resp, err := httpClient.Get(url)
 		if err != nil {
 			return err
 		}
@@ -1568,7 +1602,7 @@ func (gcp *GCP) NodePricing(key models.Key) (*models.Node, models.PricingMetadat
 		log.Debugf("Returning pricing for node %s: %+v from SKU %s", key, n.Node, n.Name)
 
 		// Add pricing URL, but redact the key (hence, "***"")
-		meta.Source = fmt.Sprintf("Downloaded pricing from %s", gcp.getBillingAPIURL("***", c.CurrencyCode))
+		meta.Source = fmt.Sprintf("Downloaded pricing from %s", gcp.buildBillingAPIURL("***", c.CurrencyCode))
 
 		n.Node.BaseCPUPrice = gcp.BaseCPUPrice
 
@@ -1588,7 +1622,7 @@ func (gcp *GCP) NodePricing(key models.Key) (*models.Node, models.PricingMetadat
 			log.Debugf("Returning pricing for node %s: %+v from SKU %s", key, n.Node, n.Name)
 
 			// Add pricing URL, but redact the key (hence, "***"")
-			meta.Source = fmt.Sprintf("Downloaded pricing from %s", gcp.getBillingAPIURL("***", c.CurrencyCode))
+			meta.Source = fmt.Sprintf("Downloaded pricing from %s", gcp.buildBillingAPIURL("***", c.CurrencyCode))
 
 			n.Node.BaseCPUPrice = gcp.BaseCPUPrice
 

--- a/pkg/cloud/gcp/provider_test.go
+++ b/pkg/cloud/gcp/provider_test.go
@@ -4,6 +4,8 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	"net/http"
+	"net/url"
 	"os"
 	"reflect"
 	"strings"
@@ -707,11 +709,62 @@ func TestGCP_findCostForDisk(t *testing.T) {
 }
 
 func TestGCP_getBillingAPIURL(t *testing.T) {
+	tests := []struct {
+		name           string
+		apiKey         string
+		currency       string
+		expectedParams map[string]string
+		absentParams   []string
+	}{
+		{
+			name:           "with API key and currency",
+			apiKey:         "test-key",
+			currency:       "USD",
+			expectedParams: map[string]string{"key": "test-key", "currencyCode": "USD"},
+		},
+		{
+			name:           "empty API key omits key param",
+			apiKey:         "",
+			currency:       "USD",
+			expectedParams: map[string]string{"currencyCode": "USD"},
+			absentParams:   []string{"key"},
+		},
+		{
+			name:           "non-USD currency",
+			apiKey:         "my-key",
+			currency:       "EUR",
+			expectedParams: map[string]string{"key": "my-key", "currencyCode": "EUR"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gcp := &GCP{}
+			query := gcp.buildBillingAPIURL(tt.apiKey, tt.currency).Query()
+
+			for param, expected := range tt.expectedParams {
+				assert.Equal(t, expected, query.Get(param), "query param %q", param)
+			}
+			for _, param := range tt.absentParams {
+				assert.False(t, query.Has(param), "query param %q should be absent", param)
+			}
+		})
+	}
+}
+
+func TestGCP_getBillingAPIClientAndURL(t *testing.T) {
 	gcp := &GCP{}
 
-	url := gcp.getBillingAPIURL("test-key", "USD")
-	expected := "https://cloudbilling.googleapis.com/v1/services/6F81-5844-456A/skus?key=test-key&currencyCode=USD"
-	assert.Equal(t, expected, url)
+	client, rawURL, err := gcp.getBillingAPIClientAndURL("test-key", "USD")
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.DefaultClient, client)
+
+	parsedURL, err := url.Parse(rawURL)
+	assert.NoError(t, err)
+	query := parsedURL.Query()
+	assert.Equal(t, "test-key", query.Get("key"))
+	assert.Equal(t, "USD", query.Get("currencyCode"))
 }
 
 func TestGCP_GpuPricing(t *testing.T) {

--- a/pkg/cloud/provider/provider.go
+++ b/pkg/cloud/provider/provider.go
@@ -143,9 +143,6 @@ func NewProvider(cache clustercache.ClusterCache, apiKey string, config *config.
 		}, nil
 	case opencost.GCPProvider:
 		log.Info("Found ProviderID starting with \"gce\", using GCP Provider")
-		if apiKey == "" {
-			return nil, errors.New("Supply a GCP Key to start getting data")
-		}
 		return &gcp.GCP{
 			Clientset:        cache,
 			APIKey:           apiKey,

--- a/pkg/cloud/provider/provider.go
+++ b/pkg/cloud/provider/provider.go
@@ -2,7 +2,6 @@ package provider
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"net"
 	"net/http"

